R/pmcmc_parameters.R

```diff
@@ -132,50 +134,142 @@ pmcmc_parameters <- R6::R6Class(
     ##' @param transform An optional transformation function to apply
     ##' to your parameter vector immediately before passing it to the
     ##' model function. If not given, then [as.list] is
-    ##' used, as dust models require this. However, if t you need to
+    ##' used, as dust models require this. However, if you need to
     ##' generate derived parameters from those being actively sampled
     ##' you can do arbitrary transformations here.
-    initialize = function(parameters, proposal, transform = NULL) {
-      parameters <- check_parameters(parameters, "pmcmc_parameter")
+    ##'
+    ##' @param populations Specifies the names of the different populations.
+    ##' Ignored if no varied parameters are included.
+    initialize = function(parameters, proposal, transform = NULL,
+                          populations = NULL) {
+      parameters <- check_parameters(parameters,
+                      c("pmcmc_parameter", "pmcmc_varied_parameter"))
 
-      if (is.null(transform)) {
-        transform <- as.list
-      }
-      assert_is(transform, "function")
+      # Split constructor for different proposal types. First asssumes array
+      #  with 3 dimensions, second is matrix.
+      varied <- vlapply(parameters, inherits, what = "pmcmc_varied_parameter")
+      if (any(varied)) {
+        # is.null(private$varied) quick check to see if all fixed
+        private$varied <- vcapply(parameters[varied], "[[", "name")
+        private$populations <- assert_character(populations)
+        lapply(parameters, function(.x) {
+          if (!(length(.x$initial) %in% c(1, length(populations)))) {
+            stop(sprintf(
+              "Expected length '1' or '%d' but got '%d' for parameter '%s'.",
+              length(populations), length(.x$initial), .x$name
+            ))
+          }
+        })
 
-      assert_is(proposal, "matrix")
-      if (!all(dim(proposal) == length(parameters))) {
-        stop(sprintf(
-          "Expected a square proposal matrix with %d rows and columns",
-          length(parameters)))
-      }
-      if (!is.null(dimnames(proposal))) {
-        ## At this point we could reorder if that's useful and these
-        ## are setequal.
-        ok <- identical(rownames(proposal), names(parameters)) &&
-          identical(colnames(proposal), names(parameters))
-        if (!ok) {
-          stop("Expected dimension names of 'proposal' to match parmeters")
+        # proposal
+        assert_is(proposal, "array")
```

Comment:
Aha, I missed the switch above on parameters being varied; that's perfect (though can you expand the docs for the constructor).

Here, we might do:

```r
if (is.matrix(proposal)) {
  ## usual checks that the proposal is ok, then
  proposal <- array(proposal, c(dim(proposal), length(populations)))
}
```

which would allow a graceful move to use the interface

Reply:
Also matrices inherit from array so `assert_is` doesn't break if proposal is `matrix` as before

Reply:
So they do. I have a feeling this changed not that long ago (perhaps for R 4.0.0?)

Reply:
I think so too